lib/pg_advisory_lock/base.rb

```diff
@@ -99,62 +99,68 @@ def with_logger
-        transaction_lock(lock_number, &block)
+        transaction_lock(lock_args, &block)
       else
-        non_transaction_lock(lock_number, &block)
+        non_transaction_lock(lock_args, &block)
       end
     end
 
-    def transaction_lock(lock_number)
+    def transaction_lock(lock_args)
       raise ArgumentError, 'block required when not within transaction' if !block_given? && !sql_caller_class.transaction_open?
 
-      return perform_lock(lock_number) unless block_given?
+      return perform_lock(lock_args) unless block_given?
 
       sql_caller_class.transaction do
-        perform_lock(lock_number)
+        perform_lock(lock_args)
         yield
       end
     end
 
-    def non_transaction_lock(lock_number)
+    def non_transaction_lock(lock_args)
       raise ArgumentError, 'block required on transaction: false' unless block_given?
 
       begin
-        perform_lock(lock_number)
+        perform_lock(lock_args)
         yield
       ensure
-        perform_unlock(lock_number)
+        perform_unlock(lock_args)
       end
     end
 
-    def perform_lock(lock_number)
+    def perform_lock(lock_args)
       function_name = "pg#{'_try' unless wait}_advisory#{'_xact' if transaction}_lock#{'_shared' if shared}"
 
       if wait
-        sql_caller_class.execute("SELECT #{function_name}(#{lock_number})")
+        sql_caller_class.execute("SELECT #{function_name}(#{lock_args})")
       else
-        result = sql_caller_class.select_value("SELECT #{function_name}(#{lock_number})")
+        result = sql_caller_class.select_value("SELECT #{function_name}(#{lock_args})")
         raise LockNotObtained, "#{self.class} can't obtain lock (#{name}, #{id.inspect})" unless result
       end
     end
 
-    def perform_unlock(lock_number)
-      sql_caller_class.select_value("SELECT pg_advisory_unlock#{'_shared' if shared}(#{lock_number})")
+    def perform_unlock(lock_args)
+      sql_caller_class.select_value("SELECT pg_advisory_unlock#{'_shared' if shared}(#{lock_args})")
     end
 
     # Converts lock name to number, because pg advisory lock functions accept only bigint numbers.
     # @return [String] lock number or two numbers delimited by comma.
-    def name_to_number
-      lock_number = _lock_names.fetch(name) do
+    def name_to_lock_args
+      lock_args = _lock_names.fetch(name) do
         raise ArgumentError, "lock name #{name.inspect} is invalid, see #{self.class}::NAMES"
       end
-      lock_number = Array.wrap(lock_number)
-      lock_number.push(id) if id.present?
-      raise ArgumentError, "can't use lock name #{name.inspect} with id" if lock_number.size > 2
+      lock_args = Array.wrap(lock_args)
+      if id.present?
+        if id.is_a?(Integer)
+          lock_args.push(id)
+        else
+          lock_args.push("(SELECT hashtext(#{ActiveRecord::Base.connection.quote(id)}))")
+        end
+      end
+      raise ArgumentError, "can't use lock name #{name.inspect} with id" if lock_args.size > 2
```

Comment:
also in `def register_lock(name, value)` method above we need to check that `value` is always integer

```ruby
raise ArgumentError, 'value must be integer' unless value.is_a?(Integer)
```